ipa-core/src/net/server/mod.rs

```diff
@@ -76,34 +76,38 @@ impl TracingSpanMaker for () {
 
 /// IPA helper web service
 ///
-/// `MpcHelperServer` handles requests from both peer helpers and external clients.
-pub struct MpcHelperServer {
-    transport: Arc<HttpTransport>,
+/// `MpcHelperServer` handles requests from peer helpers, shards within the same helper and
+/// external clients.
+///
+/// The Transport Restriction generic is used to make the server aware whether it should offer a
+/// HTTP API for shards or for other Helpers. External clients can reach out to both APIs to push
+/// the input data among other things.
+pub struct MpcHelperServer<F: ConnectionFlavor = Helper> {
     config: ServerConfig,
-    network_config: NetworkConfig,
+    network_config: NetworkConfig<F>,
+    router: Router,
 }
 
-impl MpcHelperServer {
-    pub fn new(
+impl MpcHelperServer<Helper> {
+    pub fn new_ring(
         transport: Arc<HttpTransport>,
         config: ServerConfig,
-        network_config: NetworkConfig,
+        network_config: NetworkConfig<Helper>,
     ) -> Self {
+        let router = handlers::ring_router(transport);
         MpcHelperServer {
-            transport,
             config,
             network_config,
+            router,
```

Comment:
It does not seem obvious to me why we had to to change the approach and call `self.router().clone().op()` instead of `self.router()`? can you please elaborate on that?

Reply:
If we set the router during the init we can re-use the same logic for '<Shard>' and '<Helper>' variants. 

The rest of the code, in particular `start_on`, can just assume that some Router has been set and work the same way for Shard to Shard or Helper to Helper.